### ISSUE TITLE
Keycloak v19 REGISTER event has NULL for "firstname", "lastname" after activating declarative-user-profile

### ISSUE THREAD

**Issue by marziman** [keycloak/keycloak#14431]
### Describe the bug

Hello Team,

we are facing a strange issue and we think something has changed recently to break the REGISTER event, leaving the firstname and lastname to NULL. If we create the same User in the Admin Console, the fields are not NULL. 

@thomasdarimont @pedroigor do you have an idea? This issue started after we activated the `declarative-user-profile` feature. Can this be a happening due to SPI/API changes with Quarkus Keycloak?

Many thanks for your help.
BR Mehmet

### Version

v19.0.1

### Expected behavior

REGISTER event shall contain the firstname and lastname, according the registration.

### Actual behavior

Both firstname and lastname are both NULL in the REGISTER event.
firstname = NULL
lastname = NULL

### How to Reproduce?

1. Activate declarative-user-profile at bootup
2. Set UserProfile to ON in Realm
3. Add some custom attributes, additional to the default ones (username, firstname etc.)
4. Register a User
5. Check the REGISTER events firstname, lastname fields

**Comment by pedroigor** [comment]
@marziman It should be fixed by https://github.com/keycloak/keycloak/pull/13921.

Not possible to reproduce from the nightly container image. If you can try it out, I appreciate it.

**Comment by svilenvul** [comment]
@pedroigor we have tried the nightly image, but the issue is still occurring.
Let me give you a bit more details how are we trying to access the first name and last name.

After the registration event is received in our `EventListenerProvider` implementation, we load the user information by the following code:

```
private UserModel getUserModel(KeycloakSession keycloakSession, String realmId, String userId) {
        RealmProvider model = keycloakSession.realms();
        RealmModel realm = model.getRealm(realmId);
        return keycloakSession.users().getUserById(realm, userId);
    }
```

In the case when the user profile is enabled, we cannot see the first name and last name fields. Only when we disable the user profile feature, they are visible. I am adding some sample data now to help you better understand the issue. Our event listener is producing the following data:

- with user profile enabled
```
{
  "id":"7c1d842f-6619-4bc5-854f-738da11529ce",
  "source":"/xxxx/tenant-xxxx-identity-access-management/xxxx-id-system/tiam-core-authn",
  "specversion":"1.0",
  "type":"xxxx.tiam.xxxx-id.core.authn.user.created",
  "subject":"75bb19a7-8322-42ca-a5c2-eb9bf06537b3",
  "time":"2022-09-15T15:40:03.708157Z",
  "datacontenttype":"application/json",
  "scheme":"null",
  "data":{
    "id":"75bb19a7-8322-42ca-a5c2-eb9bf06537b3",
    "name":"test@domain.de",
    "email":"test@domain.de",
    "attributes":{
      "firstName":[
        null
      ],
      "lastName":[
        null
      ],
      "email":[
        "test@domain.de"
      ],
      "username":[
        "test@domain.de"
      ]
    },
    "createdTimestamp":1663256403694,
    "isEmailVerified":false,
    "isEnabled":true
  }
}
```
- without user profile enabled
```
{
  "id":"9ae3adec-e41b-449a-a68a-3082946e64e1",
  "source":"/xxxx/tenant-xxxx-identity-access-management/xxxx-id-system/tiam-core-authn",
  "specversion":"1.0",
  "type":"xxxx.tiam.xxxx-id.core.authn.user.created",
  "subject":"80c7e052-2b2f-494f-b0b4-6e8c4875c802",
  "time":"2022-09-15T15:41:48.506347Z",
  "datacontenttype":"application/json",
  "scheme":"null",
  "data":{
    "id":"80c7e052-2b2f-494f-b0b4-6e8c4875c802",
    "name":"test@domain.de",
    "firstName":"test",
    "lastName":"domain",
    "email":"test@domain.de",
    "attributes":{
      "firstName":[
        "test"
      ],
      "lastName":[
        "domain"
      ],
      "email":[
        "test@domain.de"
      ],
      "username":[
        "test@domain.de"
      ]
    },
    "createdTimestamp":1663256508490,
    "isEmailVerified":false,
    "isEnabled":true
  }
}
```

**Comment by bs-matil** [comment]
How does your userprofile config look like?

**Comment by marziman** [comment]
We found a solution, as it turned out that the UserProfile info is sent in the KC REGISTER event before all attributes are filled out. We followed a similiar challenge others had: https://keycloak.discourse.group/t/created-user-not-immediately-available-on-event/1476

And we created a `CustomTransaction` and only send out infos after we are in `comitImpl()`

I think in general this is a timing issue, and if the KC REGISTER event would fire a bit later, all related UserProfile infos would be included. It would be great if a consumer of the KC REGISTER event can rely on getting a fully complete User with his UserProfile attributes.

Thanks!
BR Mehmet

**Comment by ElanHasson** [comment]
So this is a race condition?

Any plans to fix this?

**Comment by thomasdarimont** [comment]
I just tested this with the latest main (KC 25) where user profile is enabled by default.

When I use the default user profile and register a new user, I can see the provided firstName and lastName properly populated.
![image](https://github.com/keycloak/keycloak/assets/314690/dbcf2fc6-f05e-41d7-bef3-dcbf182ab257)

I also tested this with Keycloak 24.0.3 (also with user profile enabled by default) and I could see the firstName and lastName values in the REGISTER event.
![image](https://github.com/keycloak/keycloak/assets/314690/9e1283a6-b755-4b22-8fd8-6c95e65c9c92)

I also tested registration with a custom profile attribute:
![image](https://github.com/keycloak/keycloak/assets/314690/ac876de7-3e1b-40ec-a9b1-172f4db32b36)

![image](https://github.com/keycloak/keycloak/assets/314690/46054fed-ea17-4de6-986e-4cb5119ad80f)

So I think this works as expected in the latest versions. Am I missing something?
